contracts/rewardFunctions/IRewardFunction.sol

```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+interface IRewardFunction {
+  struct Kpi {
+    uint256 kpi;
+    address referrerAddress;
+  }
+
+  struct Reward {
+    uint256 reward;
+    address referrerAddress;
+  }
```

Comment:
should we consider matching the struct with this https://github.com/divvi-xyz/divvi-protocol/blob/main/contracts/RewardPool.sol#L24-L28? One option is both Kpi and Reward both include idempotency keys but that doesn't seem like something the RewardFunction needs to care about. We could just handle idempotency key in RewardPool contract, but probably extra data processing is required when calling calculateReward. cc @jophish

Reply:
Hmm, I think we should probably do something like have the new `RewardPool` method which accepts KPIs take (kpi, referrerAddress, idempotencyKey) tuples. We can then either pass those full tuples (including idemopotency keys) into the `calculateReward` function, and have it return (reward, referrerAddress, idempotencyKey) tuples, _or_ just pass (kpi, referrerAddress) and have the caller re-attach the idemopotency keys to the rewards. I'm not sure which option is more gas efficient. It might be less work/cleaner to do less data wrangling, however, and just always include the idempotency keys.

Reply:
Thinking more about idempotency keys, I don't think we can have 1 per kpi row, like before had 1 per reward row. When adding rewards, each row was independent, whereas now each kpi row actually affects how everyone receives rewards (and we cant just skip a row). Maybe we just need one idempotency key per reward period?